addons/website_slides/models/slide_channel.py

```diff
@@ -680,68 +781,129 @@ def action_redirect_to_members(self, completed=False):
-            for item in existing_channel_partners:
-                existing_map[item.channel_id.id].append(item.partner_id.id)
-
-            archived_members = existing_channel_partners.filtered(lambda channel_partner: not channel_partner.active)
-            if archived_members:
-                archived_members.action_unarchive()
-
-            to_create_values = [
-                dict(channel_id=channel.id, partner_id=partner.id)
-                for channel in to_join
-                for partner in target_partners if partner.id not in existing_map[channel.id]
-            ]
-            slide_partners_sudo = self.env['slide.channel.partner'].sudo().create(to_create_values)
-            to_join.message_subscribe(partner_ids=target_partners.ids, subtype_ids=[self.env.ref('website_slides.mt_channel_slide_published').id])
-            # Also consider re-joined members along with newly joined ones
-            return slide_partners_sudo | archived_members
-        return self.env['slide.channel.partner'].sudo()
+
+            # Unarchive existing channel partners
+            archived_channel_partners = existing_channel_partners.filtered(lambda channel_partner: not channel_partner.active)
+            unarchived_channel_partners_sudo = SlideChannelPartnerSudo
+            if member_status != 'invited' and archived_channel_partners:
+                archived_channel_partners.action_unarchive()
+                unarchived_channel_partners_sudo = archived_channel_partners.sudo()
+
+            existing_channel_partners_map = defaultdict(lambda: self.env['slide.channel.partner'])
+            for channel_partner in existing_channel_partners:
+                existing_channel_partners_map[channel_partner.channel_id] |= channel_partner
+
+            # Invited partners confirming their invitation by enrolling to the course.
+            channel_partners_joined_sudo = SlideChannelPartnerSudo
+            to_create_partners_values = []
+
+            for channel in channels_to_join:
+                channel_partners = existing_channel_partners_map[channel]
+                if member_status != 'invited':
+                    channel_partners_joined_sudo |= channel_partners.filtered(lambda cp: cp.member_status == 'invited')
+
+                for partner in target_partners.filtered(lambda p: p not in channel_partners.partner_id):
+                    to_create_partners_values.append(dict(channel_id=channel.id, partner_id=partner.id, member_status=member_status))
+
+            new_slide_channel_partners = SlideChannelPartnerSudo.create(to_create_partners_values)
+            channel_partners_joined_sudo.member_status = 'joined'
+
+            # Subscribe partners joining to the course to the chatter.
+            if member_status != 'invited':
+                for channel_partner in unarchived_channel_partners_sudo | channel_partners_joined_sudo | new_slide_channel_partners:
+                    channel_partner.channel_id.sudo().message_subscribe(
```

Comment:
Should be batched / channel, as if you enroll a lot of people you don' want to perform a message_subscribe / enroll.